js/modules/k6/experimental/streams/readable_stream_default_controller.go

```diff
@@ -0,0 +1,485 @@
+// [ReadableStreamDefaultController] instance.
+func NewReadableStreamDefaultControllerObject(controller *ReadableStreamDefaultController) (*goja.Object, error) {
+	rt := controller.stream.runtime
+	obj := rt.NewObject()
+
+	err := obj.DefineAccessorProperty("desiredSize", rt.ToValue(func() goja.Value {
+		desiredSize := controller.getDesiredSize()
+		if !desiredSize.Valid {
+			return goja.Null()
+		}
+		return rt.ToValue(desiredSize.Float64)
+	}), nil, goja.FLAG_FALSE, goja.FLAG_TRUE)
+	if err != nil {
+		return nil, err
+	}
+
+	// Exposing the properties of the [ReadableStreamController] interface
+	if err := setReadOnlyPropertyOf(obj, "constructor", rt.ToValue(func() goja.Value {
+		return rt.ToValue(&ReadableStreamDefaultController{})
+	})); err != nil {
+		return nil, err
+	}
+
+	if err := setReadOnlyPropertyOf(obj, "close", rt.ToValue(controller.Close)); err != nil {
+		return nil, err
+	}
+
+	if err := setReadOnlyPropertyOf(obj, "enqueue", rt.ToValue(controller.Enqueue)); err != nil {
+		return nil, err
+	}
+
+	if err := setReadOnlyPropertyOf(obj, "error", rt.ToValue(controller.Error)); err != nil {
+		return nil, err
+	}
+
+	return rt.CreateObject(obj), nil
+}
+
+// Close closes the stream.
+//
+// It implements the ReadableStreamDefaultController.close() [specification] algorithm.
+//
+// [specification]: https://streams.spec.whatwg.org/#rs-default-controller-close
+func (controller *ReadableStreamDefaultController) Close() {
+	// 1. If ! ReadableStreamDefaultControllerCanCloseOrEnqueue(this) is false, throw a TypeError exception.
+	if !controller.canCloseOrEnqueue() {
+		throw(controller.stream.vu.Runtime(), newError(TypeError, "cannot close or enqueue"))
+	}
+
+	// 2. Perform ! ReadableStreamDefaultControllerClose(this).
+	controller.close()
+}
+
+// Enqueue enqueues a chunk to the stream's internal queue.
+//
+// It implements the ReadableStreamDefaultController.enqueue(chunk) [specification] algorithm.
+//
+// [specification]: https://streams.spec.whatwg.org/#rs-default-controller-enqueue
+func (controller *ReadableStreamDefaultController) Enqueue(chunk goja.Value) {
+	// 1. If ! ReadableStreamDefaultControllerCanCloseOrEnqueue(this) is false, throw a TypeError exception.
+	if !controller.canCloseOrEnqueue() {
+		throw(controller.stream.vu.Runtime(), newError(TypeError, "cannot close or enqueue"))
+	}
+
+	// 2. Perform ? ReadableStreamDefaultControllerEnqueue(this, chunk).
+	if err := controller.enqueue(chunk); err != nil {
+		throw(controller.stream.vu.Runtime(), err)
+	}
+}
+
+// Error signals that the stream has been errored, and performs the necessary cleanup
+// steps.
+//
+// It implements the ReadableStreamDefaultController.error(e) [specification] algorithm.
+//
+// [specification]: https://streams.spec.whatwg.org/#rs-default-controller-error
+func (controller *ReadableStreamDefaultController) Error(err goja.Value) {
+	if err == nil {
+		err = goja.Undefined()
```

Comment:
This logic looks a bit weird to me; why is it so? I mean, why, in case of nil error, do we assign it to undefined and perform some `controller.error` ?

Reply:
I discovered that thanks to WPTs. I think it's because the `Error` method is a publicly exposed one, and it can be called with any argument. So, it turns that some values (probably `null`, or `undefined`, but I don't remember to be honest), is translated to `nil` in Go. However, when that error surfaces back to JS code, we need it to be a valid `goja.Value`, thus the need to sanity check it, and use a valid `goja.Value` instead.

Reply:
Quickly going over all the WPTs that call `.Error()`, I see a few calls with no arguments, so probably that was one of the cases where the Go value received is `nil`. As stated above, then JS code expects `undefined`.

Reply:
I see, okay thanks for explanation :+1: